lib/compat/wordpress-6.1/block-template-utils.php

```diff
@@ -290,3 +290,43 @@ function gutenberg_build_block_template_result_from_post( $post ) {
 	}
 	return $template;
 }
+
+/**
+ * Helper function to get the Template Hierarchy for a given slug.
+ * We need to Handle special cases here like `front-page`, `singular` and `archive` templates.
+ *
+ * Noting that we always add `index` as the last fallback template.
+ *
+ * @param string $slug The slug from which to extract the template hierarchy.
+ * @return array<string> The template hierarchy.
+ */
+function gutenberg_get_template_hierarchy( $slug ) {
+	if ( 'front-page' === $slug ) {
+		return array( 'front-page', 'home', 'index' );
+	}
+	$limit = 2;
+	if ( strpos( $slug, 'single-' ) === 0 || strpos( $slug, 'taxonomy-' ) === 0 ) {
+		// E.g. `single-post-hello` or `taxonomy-recipes-vegetarian`.
+		$limit = 3;
+	}
+	$parts = explode( '-', $slug, $limit );
+	$type  = array_shift( $parts );
+	$slugs = array( $type );
+	foreach ( $parts as $part ) {
+		array_unshift( $slugs, $slugs[0] . '-' . $part );
+	}
```

Comment:
Good catch! Maybe we should use [get_post_types](https://developer.wordpress.org/reference/functions/get_post_types/) and the respective one for taxonomies to extract the parts..

Reply:
Sounds like a great idea!

Reply:
I'll mark this resolved - see: https://github.com/WordPress/gutenberg/pull/42520#issuecomment-1195688285

We can start new discussions since this code is outdated now.